### DISCUSSION TITLE
Missing label arithmetic

### DISCUSSION THREAD

**Discussion by krisutofu** [EduMIPS64/edumips64#714] [Ideas]
Dear EduMIPS64 developers,
thank you for all your work to provide this improved educational simulator. I recently discovered EduMIPS64. I would like to switch to WinMIPS64 but I am missing a good feature that was available in WinMIPS64.

WinMIPS64 allows you to write
```
    l.d f0, array+0($zero)
    l.d f1, array+8($zero)
    l.d f2, array+16($zero)
```
It means, it will statically add 0, 8 or 16 to the immediate value that the label is substituted with.

https://github.com/EduMIPS64/edumips64/blob/4e61be815f1bdc60ba7bf5cf8198c673d3464d02/src/main/java/org/edumips64/core/parser/Parser.java#L676

The expression could support up to 5 operators `+` , `-` , `*`, `/` and `%` but should support at least `+` and `-` or specifically `label+imm` or `label-imm`.

**Comment by lupino3** [comment]
Hi @krisutofu, thanks for filing a feature request for EduMIPS64!

I am currently very behind on bugfixes and feature requests -- there is only me working on the simulator, but I'll definitely try to implement this kind of label offset arithmetic.

Our parser isn't very easy to work with, so implementing the feature will need lots of care and testing :)

In the meantime, a work-around could be to encode the offset in the register you use as the base for the addressing. For example, if you want to address `array+16`, you could do the following:

```
    daddi r1, r0, 16
    l.d f0, array($r1)
```

Hope the work-around is satisfactory in the meantime.

Thanks,
Andrea

**Comment by krisutofu** [comment]
Thank you. I wish, I could help you out, maybe a pull request but I need to find some time between university homework and work.

The workaround is functionally equivalent but a little disadvantage for homeworks where maximum performance counts.

I saw that parser story and it is very ugly, indeed. I think, breaking the code into methods would make it easier to work with. Other code places in EduMIPS64 that I have seen yet have much better code quality.

So, yes, please take your time :-) . The easiest way is only implementing `+imm` or `-imm` which would not require expression parsing (with a tree-like structure).

```
parserVal.split("[+-]");
//operands = map(substituteLabel, parserVal);
//... = reduce(add, operands);
```

More effort is probably needed for the tests, e.g.
```
ld r1, label+0(r0)
ld r1, label-0(r0)
ld r1, 0+label(r0)
ld r1, 0-label(r0)
ld r1, label+label2(r0)
ld r1, label2-label(r0)
ld r1, label-8-16(r0)
ld r1, label-8+16(r0)
//wrong:
ld r1, +label(r0)  // optional, could also be valid
ld r1, -label(r0)
ld r1, label+(r0)
ld r1, label-(r0)
ld r1, label+0+(r0)
ld r1, label+0-(r0)
ld r1, label +0(r0)  // would need further code adjustments to make an operand include whitespace
ld r1, label+ 0(r0)
ld r1, label + 0(r0)
ld r1, label - 0(r0)
ld r1, label- 0(r0)
...
```

**Reply by lupino3** [reply]
Yes, consider that we wrote most of the code when we were students and didn't have lots of experience writing well-structured code. I've tried to improve the code over the years but it's a hobby and I don't have a lot of time for it :) 

In the parser, we might need to add better validation for labels (and unit tests for validation!) before adding operators.

The sad thing is that we had written a new, more organized parser (https://github.com/EduMIPS64/edumips64/tree/new_parser/core/parser), but when we did it we didn't have unit tests so I didn't feel confident in merging it. I should probably try to find time to merge it.
